CORE-223 - Excluding CP leadership from high latency nodes (#1658)
hazelcast/hz-docs: docs/modules/cp-subsystem/pages/configuration.adoc

To scale out throughput and memory capacity, you can choose a CP group size that is
smaller than the <<cp-member-count, CP member count>> to distribute your CP data structures to multiple CP groups.

> CP groups must consist of three, five, or seven members. An odd number of CP members avoids stalemate scenarios in leadership elections and quorum calculations.

Reviewer: Must? I think it's strongly advisable but not a enforced/validated requirement.

Author: Are you sure? Current docs for `group-size` say "must".

Reviewer: Yes sorry you are correct my bad!

replicated to at least three CP members. This CP group can tolerate the failure of two CP
members and remain available.

> Creating separate CP groups allows clients to distribute the load among the leaders of several CP groups. However, a large number of CP groups running on the same members may slow down the system. A good compromise is to create the same number of CP groups as CP members. That way, each member will usually be the leader of one group.

Reviewer 2: ```suggestion
Creating separate CP groups allows clients to distribute the load among the leaders of several CP groups. However, a large number of CP groups running on the same members may slow down the system. A good compromise is to create the same number of CP groups as CP members for high throughput use cases. That way, each member will usually be the leader of one group.
```

IMPORTANT: <<cp-member-priority, `cp-member-priority`>> does not override the Raft leadership election process: it is a hint rather than a rule. Consequently, it is possible that a member with a low priority becomes leader for a period until the background leadership rebalancing task transfers the leadership to a member with a higher priority. The leadership rebalancing task runs approximately every 60 seconds; however, a custom frequency can be defined by the system property `hazelcast.raft.leadership.rebalance.period`. If necessary, you can also <<autostepdown, configure members to reject leadership>>.

[[autostepdown]]
== Configuring Members to Reject Leadership

> In some deployment scenarios, you may want to prevent a CP group member from becoming the leader; for example, if your CP group is distributed across multiple data centers and there is a high latency for communications with the remote data center(s). If the CP group leader is in the local data center -- and there are enough followers in the local data center to achieve quorum -- the leader will not need to wait for acknowledgements from followers in the remote data center and so the increased latency will have no impact. If the leader is in the remote data center, there will be a higher latency for client operations and this will reduce the maximum throughput achievable. To maximize throughput, you can prevent CP members in the remote data center from becoming leader by enabling <<auto-step-down-when-leader, 'auto-step-down-when-leader'>> on those members. In this case, the increased latency only has an impact if a remote data center member is needed to achieve quorum.

Reviewer: > If the CP group leader is in the local data center -- and there are enough followers in the local data center to achieve quorum -- the leader will not need to wait for acknowledgements from followers in the remote data center and so the increased latency will have no impact.

We want to ensure best practices are maintained in the dialogue here, this suggests that a local **single** DC achieving quorum is fine. It is not, because it impacts fault tolerance. We want to ensure we have replication of the log going across multiple DCs. 

In this example we can use a 3 DC set up. 2 DCs close together, one further away (higher latency). In this scenario, the leader ending up on the slow DC will cause higher latency.

> A member must become leader if it has the most recent log in the CP group. In this case, a member with <<auto-step-down-when-leader, 'auto-step-down-when-leader'>> enabled will become leader but immediately trigger the leadership rebalancing task, and will reject client operations. If the leadership rebalancing task fails, it will retry automatically.

Reviewer: Might be worth referencing this as a Raft property, not a hazelcast specific impl

> member must become leader if it has the most recent log in the CP group.

> Reducing the number of leadership candidates reduces the fault tolerance of the CP Subsystem. You should therefore only enable this feature if it is necessary to achieve your throughput requirements, and you should enable it on as few members as possible. This feature does not apply to the `METADATA` group, which is not sensitive to latency.

Reviewer: > , which is not sensitive to latency.

I would add,

which is required for membership operations.

Author: The METADATA group is described in cp-subsystem.adoc. Is it relevant to describe its function here?

Reviewer: I mean that the abdicating nodes still must serve the METADATA group, just an extra justification is nice for the user IMO.

On startup, any CP member with any priority can become a leader and `cp-member-priority` has no effect. The background leadership rebalancing task periodically transfers the leadership from CP members with the lowest priorities to those with higher priorities within each CP group. Eventually, all CP group leaders are CP members with high priorities.

Only CP members that are part of a CP group can become leaders. CP members that are not in a CP group do not participate in the leadership rebalancing task. For example, you start a cluster with the CP group size set to three and all members with priority 0. Later, you promote a new CP member with priority 100. Even though it has a higher priority, the new CP member never becomes a leader because the CP group is full. The new CP member is eligible for leadership only when a CP member leaves the CP group and the new CP member takes the previous member's place.

> IMPORTANT: <<cp-member-priority, `cp-member-priority`>> does not override the Raft leadership election process: it is a hint rather than a rule. Consequently, it is possible that a member with a low priority becomes leader for a period until the background leadership rebalancing task transfers the leadership to a member with a higher priority. The leadership rebalancing task runs approximately every 60 seconds; however, a custom frequency can be defined by the system property `hazelcast.raft.leadership.rebalance.period`. If necessary, you can also <<autostepdown, configure members to reject leadership>>.

Reviewer 2: ```suggestion
IMPORTANT: <<cp-member-priority, `cp-member-priority`>> does not override the Raft leadership election process: it is a hint rather than a rule. Consequently, it is possible that a member with a low priority becomes leader for a period until the background leadership rebalancing task transfers the leadership to a member with a higher priority. The leadership rebalancing task runs approximately every 60 seconds; however, a custom frequency can be defined by the system property `hazelcast.raft.leadership.rebalance.period`. If necessary, you can also <<autostepdown, configure members to automatically step down from leadership>>.
```

[[autostepdown]]
== Configuring Members to Reject Leadership

In some deployment scenarios, you may want to prevent a CP group member from becoming the leader. For example, if your CP group is distributed across two local data centers and one remote data center for geographic redundancy, there may be a significantly higher latency to the remote data center.

image:autostepdown-deployment.svg[Distributed CP Group deployment]

> If the CP group leader is in one of the local data centers -- and there are enough local followers to achieve quorum -- the leader will not need to wait for acknowledgements from followers in the remote data center and so the increased latency will have no impact. If the leader is in the remote data center, there will be a higher latency for client operations and this will reduce the maximum throughput achievable. To maximize throughput, you can prevent CP members in the remote data center from becoming leader by enabling <<auto-step-down-when-leader, `auto-step-down-when-leader`>> on those members. The increased latency will then only have an impact if a remote data center member is needed to achieve quorum.

Reviewer 2: We say "you can prevent CP members [...] becoming leader" but this isn't accurate - they do become leader for a short period of time (minimum possible) while they transfer leadership to another node.

Author: Are they the leader in any meaningful sense if they never do any of the things the leader normally does?

> In the Raft algorithm, a member must become leader if it has the most recent log in the group. In this case, a member with <<auto-step-down-when-leader, `auto-step-down-when-leader`>> enabled will become leader but immediately trigger the leadership rebalancing task, and will reject client operations. If the leadership rebalancing task fails, it will retry automatically.

Reviewer 2: ```suggestion
In the Raft algorithm, a member must become leader if it has the most recent log in the group. In this case, a member with <<auto-step-down-when-leader, `auto-step-down-when-leader`>> enabled will become leader but immediately trigger the leadership rebalancing task, and will reject client operations in the interim. If the leadership rebalancing task fails, it will retry automatically.
```

====

|===

== Per-Member Configuration Options

Use these options to configure individual CP members.

.CP member configuration options
[cols="1a,1a,1m,2a",options="header"]
|===
|Option|Description|Default|Example

|[[cp-member-priority]]`cp-member-priority`
|The priority rating as a positive or negative integer. The leader role is eventually transferred to members with higher priorities within a CP group.

|0
|
[tabs]
====
XML::
+
--
[source,xml]
----
<hazelcast>
  <cp-subsystem>
    <cp-member-priority>0
    </cp-member-priority>
  </cp-subsystem>
</hazelcast>
----
--
YAML::
+
--
[source,yaml]
----
hazelcast:
  cp-subsystem:
    cp-member-priority: 0
----
--
Java::
+
--
[source,java]
----
Config config = new Config();

config.getCPSubsystemConfig()
.setCPMemberPriority(0)

> ----

Reviewer 2: Does it make sense for our configuration example to use the default value `0`?

Author: I had the same thought. I might update the existing docs here.
